libcaf_core/caf/event_based_mail.test.cpp

```diff
@@ -0,0 +1,289 @@
+        .await([this](int value) { fail("expected a string, got: {}", value); },
+               [result](error& err) { *result = std::move(err); });
+      check_eq(mail_count(), 1u);
+      launch();
+      expect<error>().to(self_hdl);
+      check_eq(*result, make_error(sec::invalid_request));
+    }
+  }
+}
+
+TEST("send delayed request message") {
+  auto [self, launch] = sys.spawn_inactive<event_based_actor>();
+  auto self_hdl = actor_cast<actor>(self);
+  auto dummy = sys.spawn([]() -> dummy_behavior {
+    return {
+      [](int value) { return value * value; },
+    };
+  });
+  auto result = std::make_shared<int>(0);
+  auto on_result = [result](int value) { *result = value; };
+  SECTION("strong receiver reference") {
+    self->mail(3).delay(1s).request(dummy, strong_ref).first.then(on_result);
+    launch();
+    check_eq(mail_count(), 0u);
+    check_eq(num_timeouts(), 1u);
+    trigger_timeout();
+    expect<int>()
+      .with(3)
+      .priority(message_priority::normal)
+      .from(self_hdl)
+      .to(dummy);
+    expect<int>()
+      .with(9)
+      .priority(message_priority::normal)
+      .from(dummy)
+      .to(self_hdl);
+    check_eq(*result, 9);
+  }
+  SECTION("weak receiver reference") {
+    self->mail(3)
+      .schedule(sys.clock().now() + 1s)
+      .request(dummy, weak_ref)
+      .first.then(on_result);
+    launch();
+    check_eq(mail_count(), 0u);
+    check_eq(num_timeouts(), 1u);
+    trigger_timeout();
+    expect<int>()
+      .with(3)
+      .priority(message_priority::normal)
+      .from(self_hdl)
+      .to(dummy);
+    expect<int>()
+      .with(9)
+      .priority(message_priority::normal)
+      .from(dummy)
+      .to(self_hdl);
+    check_eq(*result, 9);
+  }
+  SECTION("weak receiver reference that expires") {
+    self->mail(3)
+      .schedule(sys.clock().now() + 1s)
+      .request(dummy, weak_ref)
+      .first.then(on_result);
+    launch();
+    check_eq(mail_count(), 0u);
+    check_eq(num_timeouts(), 1u);
+    dummy = nullptr;
+    trigger_timeout();
+    expect<error>().with(make_error(sec::request_receiver_down)).to(self_hdl);
+  }
+  SECTION("weak sender reference that expires") {
+    self->mail(3)
+      .schedule(sys.clock().now() + 1s)
+      .request(dummy, strong_ref, weak_self_ref)
+      .first.then(on_result);
+    launch();
+    check_eq(mail_count(), 0u);
+    check_eq(num_timeouts(), 1u);
+    self_hdl = nullptr;
+    trigger_timeout();
+    check_eq(mail_count(), 0u);
+    check_eq(num_timeouts(), 0u);
+  }
+}
+
+TEST("send request message as a typed actor") {
+  using sender_actor = typed_actor<result<void>(int)>;
+  auto dummy = sys.spawn([]() -> dummy_behavior {
+    return {
+      [](int value) { return value * value; },
+    };
+  });
+  auto [self, launch] = sys.spawn_inactive<sender_actor::impl>();
+  auto self_hdl = actor_cast<actor>(self);
+  auto result = std::make_shared<int>(0);
+  self->mail(3).request(dummy).then([result](int x) { *result = x; });
```

Comment:
If you're touching `request_handle` anyways, then I have one usability ask for it when the receiver is a typed actor:

```cpp
self->mail(3).request(...).then(
  [self](auto response) {
    // decltype(response) is expected<int>
  });
```

I know this may not be easy to implement with the current overloads available for `then(...)`, so maybe it just needs to be a separate function like `then_auto(...)`, but it would make code a lot easier to write and to maintain when using typed actors.

Reply:
Ah, didn't know that you actually had this planned now. Must've missed that in our previous discussions on this. That makes me _really_ happy! ❤️

Reply:
> Ah, didn't know that you actually had this planned now.

Sorry, that's my bad. I don't think I've ever mentioned it. 🙂

I thought about the response handle while working on this and ultimately decided splitting the response promise into two separate classes (event-based and blocking). I had that discussion from a while back still in mind and think its a worthwhile addition, but I wanted to have all the `mail` API pieces in place first.